pheweb/serve/data_access/db.py

```diff
@@ -1052,9 +1159,80 @@ def get_lofs(self, gene):
 
 class TSVDao(TSVDB):
      def __init__(self, coding):
-          self.coding_data = pd.read_csv(coding, encoding='utf8', sep='\t').fillna('NA').to_dict(orient='records')
+          df = pd.read_csv(coding, encoding='utf8', sep='\t').fillna('NA')
+          top_i = df.groupby('variant')['pval'].idxmin
+          df['is_top'] = 0
+          df.loc[top_i, 'is_top'] = 1
+          self.coding_data = df.to_dict(orient='records')
      def get_coding(self):
           return self.coding_data
+
+class FineMappingMySQLDao(FineMappingDB):
+     def __init__(self, authentication_file, base_paths):
+          self.authentication_file = authentication_file
+          self.base_paths = base_paths
+          auth_module = imp.load_source('mysql_auth', self.authentication_file)
+          self.user = getattr(auth_module, 'mysql')['user']
+          self.password = getattr(auth_module, 'mysql')['password']
+          self.host = getattr(auth_module, 'mysql')['host']
+          self.db = getattr(auth_module, 'mysql')['db']
+          self.release = getattr(auth_module, 'mysql')['release']
+     def get_connection(self):
+          return pymysql.connect(host=self.host, user=self.user, password=self.password, db=self.db)
+     def get_max_region(self, phenocode, chr, start, end):
+          conn = self.get_connection()
+          with conn.cursor(pymysql.cursors.DictCursor) as cursori:
+               sql = "SELECT min(start) as start, max(end) AS end FROM finemapped_regions WHERE rel=%s AND phenocode=%s AND chr=%s AND start <= %s AND end >= %s"
+               cursori.execute(sql, [self.release, phenocode, chr, end, start])
+               result = cursori.fetchone()
+               cursori.close()
+          conn.close()
+          return result
+     def get_regions(self, variant: Variant):
```

Comment:
type hints are great!  Let's start adding these everywhere. If you have bandwidth to do that for these changes, it would be great. IF not then we can do that later

Reply:
Agree